File: apps/node/test/chapter01/kinoko-takenoko.test.ts
import { mkdtempSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { describe, expect, it } from "vitest";
import {
  accuracy,
  loadPeople,
  predictByRule,
  splitFeaturesAndLabels,
} from "../../src/chapter01/kinoko-takenoko.ts";

const HEADER = "\uFEFF身長,体重,年代,派閥\n";

function writeCsv(rows: string): string {
  const csvFile = join(mkdtempSync(join(tmpdir(), "kvst-")), "kvst.csv");
  writeFileSync(csvFile, HEADER + rows);
  return csvFile;
}

describe("loadPeople", () => {
  it("BOM 付き CSV を読み込んで人物の配列を返す", () => {
    const csvFile = writeCsv("165,58,30,きのこ\n");

    const people = loadPeople(csvFile);

    expect(people).toEqual([
      { height: 165, weight: 58, ageGroup: 30, faction: "きのこ" },
    ]);
  });

  it("複数行の CSV を読み込んで行の順に人物の配列を返す", () => {
    const csvFile = writeCsv("161,52,20,きのこ\n183,74,50,たけのこ\n");

    const people = loadPeople(csvFile);

    expect(people).toEqual([
      { height: 161, weight: 52, ageGroup: 20, faction: "きのこ" },
      { height: 183, weight: 74, ageGroup: 50, faction: "たけのこ" },
    ]);
  });
});

describe("splitFeaturesAndLabels", () => {
  it("人物の配列を特徴量と正解ラベルに分ける", () => {
    const people = [
      { height: 161, weight: 52, ageGroup: 20, faction: "きのこ" },
      { height: 183, weight: 74, ageGroup: 50, faction: "たけのこ" },
    ];

    const { features, labels } = splitFeaturesAndLabels(people);

    expect(features).toEqual([
      { height: 161, weight: 52, ageGroup: 20 },
      { height: 183, weight: 74, ageGroup: 50 },
    ]);
    expect(labels).toEqual(["きのこ", "たけのこ"]);
  });
});

describe("predictByRule", () => {
  it("20 代ならきのこ派と判定する", () => {
    const features = { height: 161, weight: 52, ageGroup: 20 };

    expect(predictByRule(features)).toBe("きのこ");
  });

  it("20 代以外ならたけのこ派と判定する", () => {
    const features = { height: 183, weight: 74, ageGroup: 50 };

    expect(predictByRule(features)).toBe("たけのこ");
  });
});

describe("accuracy", () => {
  it("すべての予測が正解なら正解率は 1", () => {
    expect(accuracy(["きのこ", "たけのこ"], ["きのこ", "たけのこ"])).toBe(1);
  });

  it("4 件中 3 件の予測が正解なら正解率は 0.75", () => {
    const predictions = ["きのこ", "きのこ", "たけのこ", "たけのこ"];
    const labels = ["きのこ", "たけのこ", "たけのこ", "たけのこ"];

    expect(accuracy(predictions, labels)).toBe(0.75);
  });

  it("予測と正解ラベルの件数が違えばエラーになる", () => {
    expect(() => accuracy(["きのこ"], ["きのこ", "たけのこ"])).toThrow(
      "予測と正解ラベルの件数が違います",
    );
  });
});

File: apps/kotlin/src/test/kotlin/chapter01/KinokoTakenokoTest.kt
package chapter01

import dataset.dataDir
import org.junit.jupiter.api.Assumptions.assumeTrue
import support.captureStdout
import java.io.File
import java.nio.file.Path
import kotlin.io.path.createTempDirectory
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

private const val HEADER = "\uFEFF身長,体重,年代,派閥\n"

class LoadPeopleTest {
    private val directory: Path = createTempDirectory()

    private fun writeCsv(rows: String): File = File(directory.toFile(), "kvst.csv").apply { writeText(HEADER + rows) }

    @Test
    fun `BOM付きCSVを読み込んで人物のリストを返す`() {
        val csvFile = writeCsv("165,58,30,きのこ\n")

        val people = loadPeople(csvFile)

        assertEquals(listOf(Person(height = 165, weight = 58, ageGroup = 30, faction = "きのこ")), people)
    }

    @Test
    fun `複数行のCSVを読み込んで行の順に人物のリストを返す`() {
        val csvFile = writeCsv("161,52,20,きのこ\n183,74,50,たけのこ\n")

        val people = loadPeople(csvFile)

        assertEquals(
            listOf(
                Person(height = 161, weight = 52, ageGroup = 20, faction = "きのこ"),
                Person(height = 183, weight = 74, ageGroup = 50, faction = "たけのこ"),
            ),
            people,
        )
    }
}

class SplitFeaturesAndLabelsTest {
    @Test
    fun `人物のリストを特徴量と正解ラベルに分ける`() {
        val people =
            listOf(
                Person(height = 161, weight = 52, ageGroup = 20, faction = "きのこ"),
                Person(height = 183, weight = 74, ageGroup = 50, faction = "たけのこ"),
            )

        val (features, labels) = splitFeaturesAndLabels(people)

        assertEquals(
            listOf(
                Features(height = 161, weight = 52, ageGroup = 20),
                Features(height = 183, weight = 74, ageGroup = 50),
            ),
            features,
        )
        assertEquals(listOf("きのこ", "たけのこ"), labels)
    }
}

class PredictByRuleTest {
    @Test
    fun `20代ならきのこ派と判定する`() {
        val features = Features(height = 161, weight = 52, ageGroup = 20)

        assertEquals("きのこ", predictByRule(features))
    }

    @Test
    fun `20代以外ならたけのこ派と判定する`() {
        val features = Features(height = 183, weight = 74, ageGroup = 50)

        assertEquals("たけのこ", predictByRule(features))
    }
}

class AccuracyTest {
    @Test
    fun `すべての予測が正解なら正解率は1`() {
        assertEquals(1.0, accuracy(listOf("きのこ", "たけのこ"), listOf("きのこ", "たけのこ")))
    }

    @Test
    fun `4件中3件の予測が正解なら正解率は075`() {
        val predictions = listOf("きのこ", "きのこ", "たけのこ", "たけのこ")
        val labels = listOf("きのこ", "たけのこ", "たけのこ", "たけのこ")

        assertEquals(0.75, accuracy(predictions, labels))
    }

    @Test
    fun `予測と正解ラベルの件数が違えばエラーになる`() {
        assertFailsWith<IllegalArgumentException> {
            accuracy(listOf("きのこ"), listOf("きのこ", "たけのこ"))
        }
    }
}

class KvsTDataTest {
    private val csvFile = File(dataDir(), "KvsT.csv")

    @BeforeTest
    fun requireData() {
        assumeTrue(csvFile.exists(), "学習データ KvsT.csv が配置されていない（gulp data:setup）")
    }

    @Test
    fun `実データから19人分を読み込む`() {
        assertEquals(19, loadPeople(csvFile).size)
    }

    @Test
    fun `ルールによる判定の正解率を実データで計算する`() {
        val (features, labels) = splitFeaturesAndLabels(loadPeople(csvFile))

        val predictions = features.map(::predictByRule)

        assertEquals(14.0 / 19, accuracy(predictions, labels), absoluteTolerance = 1e-12)
    }

    @Test
    fun `実行するとデータ件数と正解率を表示する`() {
        val output = captureStdout { main() }

        assertEquals("データ件数: 19\nルールによる判定の正解率: 0.7368\n", output)
    }
}
